### ISSUE TITLE
Spring auto-wiring is not working with 3.2

### ISSUE THREAD

**Issue by sancar** [hazelcast/hazelcast#2251]
link to related discussion at google group

https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/hazelcast/AA3P8C7bmkI/iOS4QFcjD5gJ

**Comment by Gorgia** [comment]
looks like it's still open 3.2.2.RELEASE.
My configuration
```
<beans xmlns="http://www.springframework.org/schema/beans"
       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
       xmlns:context="http://www.springframework.org/schema/context"
       xmlns:hz="http://www.hazelcast.com/schema/spring"
       xmlns:cache="http://www.springframework.org/schema/cache"
       xsi:schemaLocation=
              "http://www.springframework.org/schema/beans
              http://www.springframework.org/schema/beans/spring-beans-3.0.xsd
              http://www.hazelcast.com/schema/spring
              http://www.hazelcast.com/schema/spring/hazelcast-spring-3.3.xsd
              http://www.springframework.org/schema/cache http://www.springframework.org/schema/cache/spring-cache.xsd">

    <bean id="applicationContextProvider" class="keyron.socialnet.ApplicationContextProvider"></bean>


    <bean id="propertyConfigurer" class="org.springframework.beans.factory.config.PropertyPlaceholderConfigurer">
        <property name="locations">
            <list>
                <value>classpath:/hazelcast-config.properties</value>
            </list>
        </property>
    </bean>

    <bean class="keyron.socialnet.hazelcast.TestBean" name="springTestBean">
        <property name="result" value="${value}"/>
    </bean>

    <hz:hazelcast id="instance">
        <hz:config>
            <hz:group name="${name}" password="${password}"/>
            <hz:properties>
                <hz:property name="hazelcast.merge.first.run.delay.seconds">5</hz:property>
                <hz:property name="hazelcast.merge.next.run.delay.seconds">5</hz:property>
            </hz:properties>
            <hz:network port="${defaultPort}" port-auto-increment="true">
                <hz:join>
                    <hz:multicast enabled="false"/>
                </hz:join>
            </hz:network>
        </hz:config>
    </hz:hazelcast>

    <hz:client id="client" lazy-init="true" scope="singleton">
        <hz:group name="${name}" password="${password}"/>
        <hz:network connection-attempt-limit="3"
                    connection-attempt-period="3000"
                    connection-timeout="1000"
                    redo-operation="true"
                    smart-routing="true">
            <hz:member>127.0.0.1:5705</hz:member>
        </hz:network>
    </hz:client>

    <hz:map             id="map"             instance-ref="instance" name="Map" lazy-init="false"/>
    <hz:multiMap        id="multiMap"        instance-ref="instance" name="Multimap"/>
    <hz:queue           id="queue"           instance-ref="instance" name="Queue"/>
    <hz:topic           id="topic"           instance-ref="instance" name="Topic" depends-on="instance, client"/>
    <hz:set             id="set"             instance-ref="instance" name="Set"/>
    <hz:list            id="list"            instance-ref="instance" name="List"/>
    <hz:executorService id="executorService" instance-ref="instance" name="ExecutorService"/>
    <hz:idGenerator     id="idGenerator"     instance-ref="instance" name="IdGenerator"/>
    <hz:atomicLong      id="atomicLong"      instance-ref="instance" name="atomicLong"/>
    <hz:atomicReference id="atomicReference" instance-ref="instance" name="atomicReference"/>
    <hz:countDownLatch  id="countDownLatch"  instance-ref="instance" name="countDownLatch"/>
    <hz:semaphore       id="semaphore"       instance-ref="instance" name="semaphore"/>
    <hz:lock            id="lock"            instance-ref="instance" name="lock"/>

</beans>

```
The ERROR

```
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'ImageGrabberTest': Injection of autowired dependencies failed; nested exception is org.springframework.beans.factory.BeanCreationException: Could not autowire field: com.hazelcast.core.HazelcastInstance ImageGrabberTest.hzclient; nested exception is org.springframework.beans.factory.NoUniqueBeanDefinitionException: No qualifying bean of type [com.hazelcast.core.HazelcastInstance] is defined: expected single matching bean but found 2: instance,client
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessPropertyValues(AutowiredAnnotationBeanPostProcessor.java:334)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1202)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireBeanProperties(AbstractAutowireCapableBeanFactory.java:385)
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.injectDependencies(DependencyInjectionTestExecutionListener.java:118)
	at org.springframework.test.context.support.DependencyInjectionTestExecutionListener.prepareTestInstance(DependencyInjectionTestExecutionListener.java:83)
	at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:212)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.createTest(SpringJUnit4ClassRunner.java:200)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner$1.runReflectiveCall(SpringJUnit4ClassRunner.java:259)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.methodBlock(SpringJUnit4ClassRunner.java:261)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:219)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:83)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:68)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:163)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:68)
Caused by: org.springframework.beans.factory.BeanCreationException: Could not autowire field: com.hazelcast.core.HazelcastInstance ImageGrabberTest.hzclient; nested exception is org.springframework.beans.factory.NoUniqueBeanDefinitionException: No qualifying bean of type [com.hazelcast.core.HazelcastInstance] is defined: expected single matching bean but found 2: instance,client
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:561)
	at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:88)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessPropertyValues(AutowiredAnnotationBeanPostProcessor.java:331)
	... 24 more
Caused by: org.springframework.beans.factory.NoUniqueBeanDefinitionException: No qualifying bean of type [com.hazelcast.core.HazelcastInstance] is defined: expected single matching bean but found 2: instance,client
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1054)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:942)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:533)
	... 26 more
```

Removing 
```
lazi-init=true 
```
remove the error. but causes
```
Exception in thread "pool-1-thread-1" com.hazelcast.core.HazelcastException: Client is shutting down!
	at com.hazelcast.client.connection.nio.ClientConnection$1.createNew(ClientConnection.java:306)
	at com.hazelcast.client.connection.nio.ClientConnection$1.createNew(ClientConnection.java:303)
	at com.hazelcast.client.connection.nio.ClientConnection.cleanResources(ClientConnection.java:349)
	at com.hazelcast.client.connection.nio.ClientConnection.innerClose(ClientConnection.java:303)
	at com.hazelcast.client.connection.nio.ClientConnection.close(ClientConnection.java:365)
	at com.hazelcast.client.connection.nio.ClientConnection.close(ClientConnection.java:231)
	at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl.shutdown(ClientConnectionManagerImpl.java:223)
	at com.hazelcast.client.impl.HazelcastClientInstanceImpl.doShutdown(HazelcastClientInstanceImpl.java:450)
	at com.hazelcast.client.impl.LifecycleServiceImpl.shutdown(LifecycleServiceImpl.java:108)
	at com.hazelcast.client.impl.HazelcastClientProxy.shutdown(HazelcastClientProxy.java:248)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.springframework.beans.factory.support.DisposableBeanAdapter.invokeCustomDestroyMethod(DisposableBeanAdapter.java:350)
	at org.springframework.beans.factory.support.DisposableBeanAdapter.destroy(DisposableBeanAdapter.java:273)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.destroyBean(DefaultSingletonBeanRegistry.java:578)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.destroySingleton(DefaultSingletonBeanRegistry.java:554)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.destroySingleton(DefaultListableBeanFactory.java:900)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.destroySingletons(DefaultSingletonBeanRegistry.java:523)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.destroySingletons(DefaultListableBeanFactory.java:907)
	at org.springframework.context.support.AbstractApplicationContext.destroyBeans(AbstractApplicationContext.java:908)
	at org.springframework.context.support.AbstractApplicationContext.doClose(AbstractApplicationContext.java:884)
	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.doClose(EmbeddedWebApplicationContext.java:150)
	at org.springframework.context.support.AbstractApplicationContext.close(AbstractApplicationContext.java:836)
	at org.springframework.boot.SpringApplication.close(SpringApplication.java:1030)
	at org.springframework.boot.SpringApplication.exit(SpringApplication.java:999)
	at keyron.socialnet.Main.main(Main.java:48)
	at ------ End remote and begin local stack-trace ------.(Unknown Source)
	at com.hazelcast.client.spi.impl.ClientCallFuture.resolveResponse(ClientCallFuture.java:202)
	at com.hazelcast.client.spi.impl.ClientCallFuture.get(ClientCallFuture.java:143)
	at com.hazelcast.client.spi.impl.ClientCallFuture.get(ClientCallFuture.java:119)
	at com.hazelcast.client.spi.ClientProxy.invokeInterruptibly(ClientProxy.java:137)
	at com.hazelcast.client.proxy.ClientQueueProxy.invokeInterruptibly(ClientQueueProxy.java:293)
	at com.hazelcast.client.proxy.ClientQueueProxy.poll(ClientQueueProxy.java:144)
	at com.hazelcast.client.proxy.ClientQueueProxy.take(ClientQueueProxy.java:139)
	at keyron.socialnet.crawler.images.ImageGrabber.run(ImageGrabber.java:45)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

Maybe due to inactivity time of bean?

**Comment by mesutcelik** [comment]
Can you create a unit test to reproduce the problem?

**Comment by Gorgia** [comment]
Unfortunately in my test case everything looks fine. Inactivity of bean may not be the problem.

**Comment by mesutcelik** [comment]
You are trying to autowire both instance and client at the same time through the same interface ```HazelcastInstance``` 

Can you try to autowire using ```@Resource``` with name attribute. ```@Resource("client")``` or ```@Resource("instance")``` ?

**Comment by Gorgia** [comment]
I think that Spring determine the autowired class also with the variable name.
This problem was due to my mistake: the main thread, which have started hazelcast and the spring context, finished before the completion of other hazelcast clients thread.
I'm still having problems with hazelcast with the queue.take method. I think that after some time, correctly hanging, hazelcast close the client's connections. Despite the fact that I set "reconnect" to true, it looks like the connection does not rebuild up fast enough, and I'm getting an exception.
However thanks for your interest.
